### PR TITLE
[Unity][Fix] Memory planning check value type of 'tir_var_upper_bound'

### PR DESCRIPTION
**Pull request by MasterJH5574**
This PR adds a sanity check to ensure that all `tir_var_upper_bound` attrs used by static memory planning has integer as value type. This check helps avoid mistakes of using wrong value types.

The check is needed since `func->GetAttr<Map<String, IntImm>>` does not apply type check.

### CODE REVIEW

File: src/relax/transform/static_plan_block_memory.cc [hunk 1]
```diff
@@ -336,6 +336,13 @@ class StorageAllocatorInit : public StorageAllocatorBaseVisitor {
     // appear in the **function signature**.
     Map<String, IntImm> var_upper_bound_attr =
         func->GetAttr<Map<String, IntImm>>("tir_var_upper_bound").value_or(Map<String, IntImm>());
+    // `GetAttr` does not perform type check.
```
Comment (junrushao) [inline src/relax/transform/static_plan_block_memory.cc:339]: “IsInstance”-style type checking, if can be determined in compile time, will be folded to a constant without actually doing the checking in runtime. Therefore, to ensure runtime type checking happens, we will need to use a step-phase downcasting, where the first phase casts an attribute to Map<ObjectRef, ObjectRef>, and then check the types of each item

**Comment by Hzfengsy** [comment]
Thanks for the fix. Beyond this commits, should we also check the `value >= 0`?

**Comment by MasterJH5574** [comment]
Thanks @junrushao and @Hzfengsy for the suggestions! Just updated the PR. Please take another look at your convenience.

**Review by junrushao — APPROVED** [review]
LGTM!

**Pull request merged**